### DISCUSSION TITLE
Example CouchDB Datasource example

### DISCUSSION THREAD

**Discussion by kellis5137** [Budibase/budibase#4487] [Help] [answered]
Hi,
I did some searching and didn't come up with anything. Is there an example of how to configure an external CouchDB data source within Budibase and what a query should look like?

Keith

**Comment by mike12345567** [comment, marked as answer]
Hi @kellis5137 - our CouchDB integration is currently somewhat limited compared to say our SQL integrations - in general it makes use of the [PouchDB API](https://pouchdb.com/api.html).

To connect to your Couch database you'll need to provide a connection string in the format of `http://<username>:<password>@<url>:<port>` e.g. `http://user:pass@localhost:5984`. Currently the integration only allows connecting to a single Couch database per datasource.

The query functions map to the PouchDB API in the following way:
1. Read - this is an `allDocs` call, so in the query you could do something like `{ "keys": [ "doc_id" ] }` and it would return that document, you can also use say `start_key`/`end_key` to get a series of docs.
2. Create - this maps to a `post` call, so everything put into the query will be written as a document, assuming no `_id` field is provided.
3. Update - this maps to a `put` call, very similar to the creation function but an `_id` and `_rev` are required to update the document.
4. Delete - maps to the `remove` call.

Currently views, search indexes and mango queries are not supported, however if there is interest there is a possibility that CouchDB could be added to the roadmap as a new datasource plus, meaning improved integration with Budibase datasources.
